### PR TITLE
Aether merge and Esper update

### PR DESCRIPTION
**Pull request by AndrewJakubowicz**
Working progress of integrating Aether into our codebase and upgrading esper to use esper language plugins.

Completed:

- Aether merge working
  - JavaScript language working in verifier.
- Esper plugins loading via script tag.
- Did full QA on change.

Incomplete:

- ~~Esper python plugin throwing error for python, Dungeons of Kithgard: `TypeError: class constructors must be invoked with |new|`.~~
- Aether tests must run through karma.

### CODE REVIEW

**Comment by AndrewJakubowicz** [comment]
Slowly trying to move in tests and fix errors as they present themselves.

Moved `aether_spec` into the test folder. Needed to include `esper.moden.js` in the karma config to fix `esper.plugin is not a function` errors.

Now I started working on moving `constructor_spec.coffee` into the test folder. When running via `localhost:3000/test` there is this concerning error:

`JSHint died with error TypeError: "_.includes is not a function"`

`_.includes` is a lodash 4 method. I think this may be the inverse of the lodash silently upgrading to version 4 when we require `aether.coffee` instead of the build artifact.

**Comment by AndrewJakubowicz** [comment]
Progress update.

JsHint has been disabled as it breaks everything and isn't critical.

There is a cryptic error that is caused in the python plugin for the python tests. This error is a mismatch of the plugin modern version to the engine modern version.

It can be fixed when python-modern plugin is loaded with esper.modern. By default the python plugin was not the modern version.

TODO:
Thus the esper language loading system has to be updated to load the language plugins based on what esper is loaded.

**Comment by AndrewJakubowicz** [comment]
Language plugins load correctly based on the browser.
I tested intro:37 and course-2:72 in the verifier for both python and javascript and they passed.
The tests that came with the merge are still partially broken. The errors that esper provides is overwriting some of our aether hacked errors and thus breaking tests.

Next steps:
Find out why game dev is throwing this error:

```
|Poseidon's Lana| Non-UserCodeError: TypeError: match.score is not a function
TypeError: match.score is not a function
    at Thang.PropertyErrorHelper.getScoreMatch (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:122:26)
    at Thang.PropertyErrorHelper.checkForProbableError (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:79:19)
    at Object.set (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:24:26)
    at Proxy.Moves.calculateRollingResistance (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:324:42)
    at Movement.updateMotion (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:77:17)
    at Movement.update (eval at module.exports.World.loadClassFromCode (webpack-internal:///48), <anonymous>:36:14)
    at WorldFrame.module.exports.WorldFrame.getNextFrame (webpack-internal:///39:28:27)
    at World.module.exports.World.getFrame (webpack-internal:///48:125:38)
    at World.module.exports.World.loadFrames (webpack-internal:///48:256:14)
    at runWorld (http://localhost:3000/javascripts/workers/worker_world.js:422:14)
    at http://localhost:3000/javascripts/workers/worker_world.js:577:24
```

**Comment by AndrewJakubowicz** [comment]
Found out that match.score was an error due to a library that we expected to mutate the string prototype.
This was resolved in this commit [c8c0f3c](https://github.com/codecombat/codecombat/pull/5080/commits/c8c0f3c7f4417800348d04021e0997286467ec5b).

File: app/assets/javascripts/workers/aether_worker.js [hunk 1]
```diff
@@ -19,8 +19,16 @@ var languagesImported = {};
 
 var ensureLanguageImported = function(language) {
   if (languagesImported[language]) return;
-  if (language === 'html') return;
-  importScripts("/javascripts/app/vendor/aether-" + language + ".js");
+  if (language === 'html' || language === 'javascript') return;
```
Comment (AndrewJakubowicz) [inline app/assets/javascripts/workers/aether_worker.js]: Here we're loading in the language plugin. JavaScript is build into esper engine.

File: app/assets/javascripts/workers/aether_worker.js [hunk 2]
```diff
@@ -19,8 +19,16 @@ var languagesImported = {};
...
+  //Detect very modern javascript support.
+  try {
```
Comment (AndrewJakubowicz) [inline app/assets/javascripts/workers/aether_worker.js]: Similar check to the esper engine to ensure the right plugin version chosen for the esper engine.

File: app/assets/javascripts/workers/worker_world.js [hunk 3]
```diff
@@ -82,7 +82,15 @@ var languagesImported = {};
 var ensureLanguageImported = function(language) {
   if (languagesImported[language]) return;
   if (language === 'javascript' || language === 'html') return;  // Only has JSHint, but we don't need to lint here.
-  myImportScripts("/javascripts/app/vendor/aether-" + language + ".js");
+  //Detect very modern javascript support.
+  try {
```
Comment (AndrewJakubowicz) [inline app/assets/javascripts/workers/worker_world.js]: Again plugin version loading. Code is repeated as it's in the workers.

File: app/lib/aether/aether.coffee [hunk 4]
```diff
@@ -0,0 +1,244 @@
+self = window if window? and not self?
```
Comment (AndrewJakubowicz) [inline app/lib/aether/aether.coffee]: This file is the main aether file that has been copy pasted in.

Comment (AndrewJakubowicz) [reply app/lib/aether/aether.coffee]: It is hard to know which of the public interface is still being used. So currently there is probably lots of excess code.

File: app/lib/aether/aether.coffee [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+  canTranspile: (rawCode, thorough=false) ->
+    return true if not rawCode # blank code should compile, but bypass the other steps
+    return false if @language.obviouslyCannotTranspile rawCode
+    return true unless thorough
+    @lint(rawCode, @).errors.length is 0
+
+  # Determine whether two strings of code are significantly different.
+  # If careAboutLineNumbers, we strip trailing comments and whitespace and compare line count.
+  # If careAboutLint, we also lint and make sure lint problems are the same.
+  hasChangedSignificantly: (a, b, careAboutLineNumbers=false, careAboutLint=false) ->
```
Comment (AndrewJakubowicz) [inline app/lib/aether/aether.coffee]: This method is called a lot however the implementation for various languages doesn't actually work. So it always defaults to `true`, I believe.

File: app/lib/aether/aether.coffee [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+    aLintProblems = ([p.id, p.message, p.hint] for p in @getAllProblems @lint a)
+    bLintProblems = ([p.id, p.message, p.hint] for p in @getAllProblems @lint b)
+    return not _.isEqual aLintProblems, bLintProblems
+
+  # Return a beautified representation of the code (cleaning up indentation, etc.)
+  beautify: (rawCode) ->
+    @language.beautify rawCode, @
+
+  # Transpile it. Even if it can't transpile, it will give syntax errors and warnings and such. Clears any old state.
+  transpile: (@raw) ->
```
Comment (AndrewJakubowicz) [inline app/lib/aether/aether.coffee]: This is a very important method. This one is called often to generate the abstract syntax tree in the `purifyCode` method.

File: app/lib/aether/aether.coffee [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+    @pure
+
+  # Perform some fast static analysis (without transpiling) and find any lint problems.
+  lint: (rawCode) ->
+    lintProblems = errors: [], warnings: [], infos: []
+    @addProblem problem, lintProblems for problem in @language.lint rawCode, @
+    lintProblems
+
+  # Return a ready-to-interpret function from the parsed code.
+  createFunction: ->
```
Comment (AndrewJakubowicz) [inline app/lib/aether/aether.coffee]: This is a method that is called from the Level Editor a lot. I don't understand it fully yet. However it appears to use esper within it.

File: app/lib/aether/aether.coffee [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+  addProblem: (problem, problems=null) ->
+    return if problem.level is "ignore"
+    (problems ? @problems)[problem.level + "s"].push problem
+    problem
+
+  # Return all the problems as a flat array.
+  getAllProblems: (problems) ->
+    _.flatten _.values (problems ? @problems)
+
+  ###
```
Comment (AndrewJakubowicz) [inline app/lib/aether/aether.coffee]: Added these comments as this appears to be quite an important method.

File: app/lib/aether/interpreter.coffee [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+              variables[n] = p.value.debugString if p.value
+          f.variables = variables
+
+        rng = top.ast.originalRange
+        f.range = [rng.start, rng.end] if rng
+        f.type = top.ast.type
+
+        bottom.flow.statements.push f unless not f.range # Dont push statements without ranges
+
+###
```
Comment (AndrewJakubowicz) [inline app/lib/aether/interpreter.coffee]: This is another method I have modified to make coffeescript work.

File: app/lib/aether/interpreter.coffee [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+  esper.languages.coffeescript.parser("# Rawr", {inFunctionBody: true})
+  ```
+###
+module.exports.parse = (aether, code) ->
+  esper = window?.esper ? self?.esper ? global?.esper ? require 'esper.js'
+  esper.plugin 'lang-' + aether.language.id
+  return esper.languages[aether.language.id].parser(code) if aether.language.id in ["coffeescript"]
+  return esper.languages[aether.language.id].parser(code, inFunctionBody: true)
+
+module.exports.createFunction = (aether) ->
```
Comment (AndrewJakubowicz) [inline app/lib/aether/interpreter.coffee]: This is another function we use that I haven't properly understood.

File: app/lib/aether/languages/coffeescript.coffee [hunk 11]
```diff
@@ -0,0 +1,17 @@
+_ = window?._ ? self?._ ? global?._ ? require 'lodash'  # rely on lodash existing, since it busts CodeCombat to browserify it--TODO
```
Comment (AndrewJakubowicz) [inline app/lib/aether/languages/coffeescript.coffee]: These languages don't seem to exist and I should probably try to delete them.
The issue is that the method `useFunctionWrapping` is required in the main code flow even if the rest isn't.

File: app/lib/aether/aether.coffee [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+    varNames = {}
+    varNames[parameter] = true for parameter in @options.functionParameters
+    preNormalizationTransforms = [
+      transforms.makeGatherNodeRanges originalNodeRanges, wrappedCode, @language.wrappedCodePrefix
+      transforms.makeCheckThisKeywords @allGlobals, varNames, @language, @options.problemContext
+      transforms.makeCheckIncompleteMembers @language, @options.problemContext
+    ]
+    try
+      #[transformedCode, transformedAST] = @transform wrappedCode, preNormalizationTransforms, @language.parse
+      @ast = interpreter.parse @, wrappedCode
```
Comment (differentmatt) [inline app/lib/aether/aether.coffee]: How does the interpreter know which language to parse?

File: app/lib/aether/interpreter.coffee [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+      else if aether.options.yieldAutomatically
+        fx = engine.functionFromAST aether.ast, (engine) -> true
+      else
+        fx = engine.functionFromASTSync aether.ast
+  catch error
+    console.log 'Esper: error parsing AST. Returning empty function.', error.message
+    if aether.language.id is 'javascript'
+      error.message = "Couldn't understand your code. Are your { and } braces matched?"
+    else
+      error.message = "Couldn't understand your code. Do you have extra spaces at the beginning, or unmatched ( and ) parentheses?"
```
Comment (differentmatt) [inline app/lib/aether/interpreter.coffee]: Are these error messages displayed to the user?  How are they localized?

**Comment by differentmatt** [comment]
What is the summary of this work?  E.g. why are we merging Aether, what does that mean, why are we adding esper plugin support, what does that mean.

Is there any easy way to see what has changed vs. been copied over from Aether repo?
Did all of the Aether tests get copied over as well?  Are they working?
How do we know the user code error and warning ranges are still working correctly?

**Comment by AndrewJakubowicz** [comment]
Summary of this work:
We are merging Aether because it's relied upon less and less. There's an end goal where we have completely moved away from using Aether and use Esper entirely instead.
By moving Aether into our codebase we are removing some indirection between the game client -> aether -> esper.

This merge also updates esper to use esper plugins for the various languages.

Updated esper provides better error messages as well as better performance.

It is not easy to see what has been changed from the original Aether repo unless you compare it yourself. In the majority of cases there is less here as esper allows us to use less of Aether's functionality.

The Aether tests have been copied over but are not working. They don't work as the `JSHint` linting library breaks in our global lodash environment when running alongside other tests. It's unknown why it works in the game.

Future work is to get new tests running.

We have a fairly large confidence that warnings and errors work through comprehensive QA testing.
This is still a chance for new bugs to be introduced. The new Esper engine is more robust in catching errors and I've observed better error messages for cases we haven't handled before.

**Pull request merged**